File: internal/interface/preferences_test.go
// Copyright 2026 Stéphane Primault <sprimault@users.noreply.github.com>
// SPDX-License-Identifier: Apache-2.0

package ihm

import (
	"net/http"
	"strings"
	"testing"
	"testing/fstest"

	"github.com/sprimault/ormeau/internal/config"
)

// pageServie rend le HTML de la page d'entrée.
func pageServie(t *testing.T, s *serveur, routeur http.Handler) string {
	t.Helper()

	w := lire(s, routeur, "/")
	if w.Code != http.StatusOK {
		t.Fatalf("code %d, attendu %d", w.Code, http.StatusOK)
	}
	return w.Body.String()
}

// TestPageInjecteLesPreferences vérifie que thème, langue et tailles arrivent
// dans la page elle-même.
//
// C'est tout l'objet du lot : le port étant tiré à chaque lancement, le
// navigateur ne peut rien retenir, et une valeur qui arriverait par un appel
// d'API arriverait après le premier rendu, donc après la bascule qu'on veut
// supprimer.
func TestPageInjecteLesPreferences(t *testing.T) {
	t.Parallel()

	s, routeur := serveurDeTest(t)
	ouvert := false
	if err := s.preferences.ecrire(config.Preferences{
		Theme:         "sombre",
		Langue:        "en",
		ApercuOuvert:  &ouvert,
		LargeurArbre:  320,
		HauteurApercu: 240,
	}); err != nil {
		t.Fatalf("écriture : %v", err)
	}

	page := pageServie(t, s, routeur)
	for _, attendu := range []string{
		`lang="en"`,
		`data-theme="sombre"`,
		`data-apercu-ouvert="0"`,
		"--ormeau-largeur-arbre:320px",
		"--ormeau-hauteur-apercu:240px",
	} {
		if !strings.Contains(page, attendu) {
			t.Errorf("la page ne porte pas %s", attendu)
		}
	}
	// Jamais réglée : sa variable n'est pas écrite et le composant garde son
	// défaut, plutôt que de recevoir un zéro qui écraserait la mise en page.
	if strings.Contains(page, "--ormeau-largeur-entites") {
		t.Error("une taille non réglée écrit quand même sa variable")
	}
}

// TestPageSansPreferencesPorteLesDefauts couvre le premier lancement.
func TestPageSansPreferencesPorteLesDefauts(t *testing.T) {
	t.Parallel()

	s, routeur := serveurDeTest(t)

	page := pageServie(t, s, routeur)
	if !strings.Contains(page, `lang="fr"`) || !strings.Contains(page, `data-theme="systeme"`) {
		t.Error("la page ne porte pas les valeurs par défaut")
	}
	if strings.Contains(page, "data-apercu-ouvert") {
		t.Error("un aperçu jamais réglé écrit quand même son attribut")
	}
}

// TestPreferencesLuesEtEcrites vérifie l'aller-retour par l'API, qui sert aux
// relectures et aux réglages, jamais au premier rendu.
func TestPreferencesLuesEtEcrites(t *testing.T) {
	t.Parallel()

	s, routeur := serveurDeTest(t)

	initiales := decoderReponse[config.Preferences](t, lire(s, routeur, "/api/preferences"))
	if initiales.Theme != "systeme" || initiales.Langue != "fr" {
		t.Errorf("défauts %+v", initiales)
	}

	w := poster(s, routeur, "/api/preferences", config.Preferences{Theme: "clair", Langue: "en", LargeurArbre: 300})
	attendreStatut(t, w, http.StatusOK)

	relues := decoderReponse[config.Preferences](t, lire(s, routeur, "/api/preferences"))
	if relues.Theme != "clair" || relues.Langue != "en" || relues.LargeurArbre != 300 {
		t.Errorf("préférences relues %+v", relues)
	}
}

// TestPreferencesRefuseUneValeurInconnue vérifie que la validation du paquet
// config s'applique aussi à ce qui vient du navigateur.
func TestPreferencesRefuseUneValeurInconnue(t *testing.T) {
	t.Parallel()

	s, routeur := serveurDeTest(t)

	w := poster(s, routeur, "/api/preferences", config.Preferences{Theme: "fluo", Langue: "fr"})
	attendreStatut(t, w, http.StatusUnprocessableEntity)

	inchangees := decoderReponse[config.Preferences](t, lire(s, routeur, "/api/preferences"))
	if inchangees.Theme != "systeme" {
		t.Errorf("un refus a quand même changé l'état : %+v", inchangees)
	}
}

// TestIndexSansAncreRefuseDeDemarrer couvre le jour où la page produite par
// Vite ne portera plus le point d'insertion.
//
// Un démarrage qui échoue en le nommant, plutôt qu'une page servie telle
// quelle : le thème ne tiendrait plus, et rien ne rattacherait le symptôme à ce
// changement.
func TestIndexSansAncreRefuseDeDemarrer(t *testing.T) {
	t.Parallel()

	cas := map[string]string{
		"balise html modifiée": `<!doctype html><html lang="en"></head>`,
		"tête absente":         `<!doctype html><html lang="fr">`,
	}
	for nom, page := range cas {
		t.Run(nom, func(t *testing.T) {
			t.Parallel()

			_, err := indexTemplate(fstest.MapFS{"index.html": {Data: []byte(page)}})
			if err == nil {
				t.Fatal("index sans ancre accepté")
			}
			if !strings.Contains(err.Error(), "ancre") {
				t.Errorf("l'erreur ne dit pas ce qui manque : %v", err)
			}
		})
	}
}

File: web/src/shared/model/__tests__/preferences.test.ts
// Copyright 2026 Stéphane Primault <sprimault@users.noreply.github.com>
// SPDX-License-Identifier: Apache-2.0

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import type { Preferences } from '../config';

/**
 * Charge le module après avoir préparé le document.
 *
 * L'état initial est lu au chargement, comme dans le navigateur où la page
 * arrive déjà injectée : un store importé en tête de fichier aurait figé les
 * valeurs avant que le test ne pose quoi que ce soit.
 */
async function chargerStore() {
  vi.resetModules();
  return await import('../preferences');
}

/** Pose ce que le serveur écrit dans la page. */
function injecter(attributs: Record<string, string>, variables: Record<string, string> = {}) {
  const racine = document.documentElement;
  for (const [nom, valeur] of Object.entries(attributs)) {
    racine.setAttribute(nom, valeur);
  }
  for (const [nom, valeur] of Object.entries(variables)) {
    racine.style.setProperty(nom, valeur);
  }
}

describe('préférences', () => {
  beforeEach(() => {
    const racine = document.documentElement;
    for (const nom of ['data-theme', 'data-apercu-ouvert']) {
      racine.removeAttribute(nom);
    }
    racine.removeAttribute('lang');
    racine.style.cssText = '';
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('lit ce que le serveur a injecté dans la page', async () => {
    injecter(
      { 'data-theme': 'sombre', lang: 'en', 'data-apercu-ouvert': '0' },
      { '--ormeau-largeur-arbre': '320px', '--ormeau-hauteur-apercu': '240px' },
    );

    const { usePreferencesStore } = await chargerStore();
    const preferences = usePreferencesStore.getState().preferences;

    expect(preferences.theme).toBe('sombre');
    expect(preferences.langue).toBe('en');
    expect(preferences.apercu_ouvert).toBe(false);
    expect(preferences.largeur_arbre).toBe(320);
    expect(preferences.hauteur_apercu).toBe(240);
    expect(preferences.largeur_entites).toBeUndefined();
  });

  it('retombe sur ses défauts quand la page n’est pas injectée', async () => {
    const { usePreferencesStore } = await chargerStore();
    const preferences = usePreferencesStore.getState().preferences;

    // Le cas de `make dev`, où Vite sert la page sans passer par le serveur Go.
    expect(preferences.theme).toBe('systeme');
    expect(preferences.langue).toBe('fr');
    expect(preferences.apercu_ouvert).toBeUndefined();
  });

  it('n’enregistre qu’une fois une rafale de réglages', async () => {
    vi.useFakeTimers();
    const { usePreferencesStore, brancherEnregistrement } = await chargerStore();

    const enregistrees: Preferences[] = [];
    brancherEnregistrement((p) => enregistrees.push(p));

    // Ce que produit une poignée qu'on tire : un événement par image.
    for (const largeur of [300, 320, 340, 360]) {
      usePreferencesStore.getState().regler({ largeur_arbre: largeur });
    }
    expect(enregistrees).toHaveLength(0);

    vi.runAllTimers();
    expect(enregistrees).toHaveLength(1);
    expect(enregistrees[0].largeur_arbre).toBe(360);
  });

  it('applique le réglage sans attendre l’enregistrement', async () => {
    vi.useFakeTimers();
    const { usePreferencesStore } = await chargerStore();

    usePreferencesStore.getState().regler({ theme: 'clair' });
    expect(usePreferencesStore.getState().preferences.theme).toBe('clair');
  });

  it('nomme la variable CSS de chaque taille', async () => {
    const { variableCSS } = await chargerStore();

    expect(variableCSS('largeur_arbre')).toBe('--ormeau-largeur-arbre');
    expect(variableCSS('hauteur_apercu')).toBe('--ormeau-hauteur-apercu');
    expect(variableCSS('largeur_entites')).toBe('--ormeau-largeur-entites');
  });
});
